### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by virtushawk**
Pull request for module 2

### CODE REVIEW

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            " FROM gift_certificate WHERE id = ?";
+    private static final String SQL_INSERT_CERTIFICATE = "INSERT INTO gift_certificate(name,description,price,duration,create_date,last_update_date) " +
+            "VALUES(?,?,?,?,?,?)";
+    private static final String SQL_DELETE_CERTIFICATE_BY_ID = "DELETE FROM gift_certificate WHERE id = ?";
+    private static final String SQL_DELETE_TAG_HAS_GIFT_CERTIFICATE_BY_ID = "DELETE FROM tag_has_gift_certificate WHERE gift_certificate_id = ?";
+    private static final String SQL_UPDATE_BY_ID = "UPDATE gift_certificate SET %s = ? WHERE id = ?";
+    private static final String SQL_UPDATE_CERTIFICATE = "UPDATE gift_certificate SET name = ?, description = ?, price = ?, " +
+            "duration = ?, last_update_date = ? WHERE id = ?";
+    private static final String LAST_UPDATE_DATE_COLUMN = "last_update_date";
+    private static final String SELECT_CERTIFICATE_QUERY = "SELECT gift_certificate.id,gift_certificate.name,gift_certificate.description,gift_certificate.price,gift_certificate.duration,gift_certificate.create_date,gift_certificate.last_update_date FROM gift_certificate" +
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java:56]: Please remove unused code

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+/**
+ * The type Sql statement formatter.
+ */
+@UtilityClass
+public class SQLStatementFormatter {
+
+    private static final String SELECT_CERTIFICATE_QUERY = "SELECT gift_certificate.id,gift_certificate.name,gift_certificate.description,gift_certificate.price,gift_certificate.duration,gift_certificate.create_date,gift_certificate.last_update_date FROM gift_certificate" +
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: Not good idea to save SQL queries in different places

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            " WHERE (:tag IS NULL OR tag.name = :tag) AND (:text IS NULL OR (gift_certificate.name LIKE %s OR gift_certificate.description LIKE %s)) " +
+            "ORDER BY %s";
+
+    /**
+     * Prepare certificate query statement string.
+     *
+     * @param querySpecification the query specification
+     * @return the string
+     */
+    public static String prepareCertificateQueryStatement(QuerySpecification querySpecification) {
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: Do we need an utility class for this method? Could we move it to CertificateDAOImpl as private method? As I can see we use it only there

File: controller/build.gradle [hunk 4]
```diff
@@ -0,0 +1,21 @@
+plugins {
+    id 'war'
+}
+
+repositories {
+    mavenCentral()
```
Comment (maximelyashevich) [inline controller/build.gradle]: Can we create parent `build.gradle` like this?

```groovy
plugins {
   id 'java'
}
 
allprojects {
   plugins.withType(JavaPlugin) {
        dependencies {
             // implementation...
        }

        test {
             useJUnitPlatform()
        }

        repositories {
             mavenCentral()
        }
    }
}
```

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * The type Certificate dao.
+ */
+@Repository
+@AllArgsConstructor
+@Slf4j
+public class CertificateDAOImpl implements CertificateDAO {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java:39]: please add free space between lines

File: controller/src/main/java/com/epam/esm/config/WebConfig.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+@ComponentScan("com.epam.esm")
+public class WebConfig implements WebMvcConfigurer {
+
+    @Override
+    public void configureDefaultServletHandling(DefaultServletHandlerConfigurer configurer) {
+        configurer.enable();
+    }
+
+    /**
+     * Init binder.
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/config/WebConfig.java]: Please make it more clear and review `javadoc` for this class.

File: controller/src/main/java/com/epam/esm/config/WebConfig.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.support.ResourceBundleMessageSource;
+import org.springframework.web.bind.WebDataBinder;
+import org.springframework.web.bind.annotation.InitBinder;
+import org.springframework.web.servlet.config.annotation.DefaultServletHandlerConfigurer;
+import org.springframework.web.servlet.config.annotation.EnableWebMvc;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+/**
+ * The type Web config.
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/config/WebConfig.java]:
```suggestion
 * The type Web configuration.
```

File: controller/src/main/java/com/epam/esm/config/DispatcherServletConfig.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package com.epam.esm.config;
+
+import org.springframework.web.servlet.support.AbstractAnnotationConfigDispatcherServletInitializer;
+
+/**
+ * The type Dispatcher servlet config.
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/config/DispatcherServletConfig.java]:
```suggestion
 * The type Dispatcher servlet configuration.
```

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+                int i = 0;
+                for (String sort : querySpecification.getSort()) {
+                    sorts.set(i, sort);
+                    i++;
+                    if (i == sorts.size()) {
+                        break;
+                    }
+                }
+            }
+            result = new StringBuilder();
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: init variable one more time? Please think how to refactor it.

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+            int i = 0;
+            for (String order : querySpecification.getOrder()) {
+                result.append(order).append(' ').append(sorts.get(i).toUpperCase(Locale.ROOT)).append(',');
+            }
+            result.deleteCharAt(result.length() - 1);
+        }
+        return String.format(SELECT_CERTIFICATE_QUERY, text, text, result);
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: please use informative variable names

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+import com.epam.esm.dto.CertificateDTO;
+import com.epam.esm.dto.PatchDTO;
+import com.epam.esm.dto.QuerySpecificationDTO;
+import com.epam.esm.exception.IdInvalidException;
+import com.epam.esm.exception.InvalidDataFormException;
+import com.epam.esm.service.CertificateService;
+import lombok.AllArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.*;
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CertificateController.java]: Please replace it with single import statements, same for `TagController`, `CertificateDAOImpl`, `TestConfig`, `CertificateDTO`, `PatchDTO`, `CertificateServiceImpl`, `CertificateServiceImplTest` classes.

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+import javax.validation.Valid;
+import java.util.List;
+
+/**
+ * The type Certificate controller.
+ */
+@RequestMapping(value = "/certificates")
+@RestController
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CertificateController.java]:
```suggestion
@RestController
@RequestMapping(value = "/certificates")
```

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    private final CertificateService service;
+
+    /**
+     * Gets all certificates.
+     *
+     * @param querySpecificationDTO the query specification dto
+     * @return List of all certificates
+     */
+    @GetMapping
+    public List<CertificateDTO> getAllCertificates(QuerySpecificationDTO querySpecificationDTO) {
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CertificateController.java]: I would suggest using `findAllCertificatesByCriteria` for this method and `findCertificateById` - for `getCertificate` method. What do you think?

File: persistence/src/main/java/com/epam/esm/mapper/CertificateMapper.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * The type Certificate mapper.
+ */
+@Component
+public class CertificateMapper implements RowMapper<Certificate> {
+
+    @Override
+    public Certificate mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return Certificate.builder().id(rs.getLong("id")).name(rs.getString("name"))
```
Comment (ykazlovich) [inline persistence/src/main/java/com/epam/esm/mapper/CertificateMapper.java]: I suppose a new line for an each field will be looked better

File: persistence/src/test/java/com/epam/esm/dao/TagDAOImplTest.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Test
+    void findByIdValid() {
+        Long id = 1L;
+        Optional<Tag> actual = tagDAO.findById(id);
+        Assertions.assertEquals("IT", actual.get().getName());
+    }
+
+    @Test
+    void findByIdEmpty() {
```
Comment (ykazlovich) [inline persistence/src/test/java/com/epam/esm/dao/TagDAOImplTest.java]: As I understood we try to find not existing tag. The name should demonstrate it

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * Create certificate
+     *
+     * @param certificateDTO the certificate dto
+     * @param bindingResult  the binding result
+     * @return the certificate dto
+     */
+    @PostMapping
+    public CertificateDTO createCertificate(@Valid @RequestBody CertificateDTO certificateDTO, BindingResult bindingResult) {
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CertificateController.java]: What's about `HttpStatus.CREATED`  for this and `TagController` class?

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    @PatchMapping(value = "/{id}")
+    public CertificateDTO patchCertificate(@PathVariable Long id, @Valid @RequestBody PatchDTO patchDTO, BindingResult bindingResult) {
+        if (id < 0) {
+            throw new IdInvalidException(id);
+        }
+        if (bindingResult.hasErrors()) {
+            throw new InvalidDataFormException(bindingResult);
+        }
+        patchDTO.setId(id);
+        return service.applyPatch(patchDTO);
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CertificateController.java]: As an alternative, you can use `service.applyPatch(id, patchDTO)` method.

File: controller/src/main/java/com/epam/esm/controller/ExceptionController.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+import java.util.Locale;
+
+/**
+ * The type Exception controller.
+ */
+@ControllerAdvice
+@ResponseBody
+public class ExceptionController extends ResponseEntityExceptionHandler {
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/ExceptionController.java]: Just an idea:
```suggestion
public class CustomExceptionHandler extends ResponseEntityExceptionHandler {
```

File: controller/src/main/java/com/epam/esm/controller/ExceptionController.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Autowired
+    private MessageSource messageSource;
+
+    private static final int INTERNAL_SERVER_ERROR_CODE = 100;
+
+    /**
+     * Handle controller not found exception response entity.
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/ExceptionController.java]: Please review `javadoc` for this class.

File: controller/src/main/java/com/epam/esm/controller/ExceptionController.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+     * @param exception the exception
+     * @param locale    the locale
+     * @return the response entity
+     */
+    @ExceptionHandler(CertificateNotFoundException.class)
+    public ResponseEntity<ErrorResponse> handleControllerNotFoundException(CertificateNotFoundException exception, Locale locale) {
+        String errorMessage = messageSource.getMessage("error.certificateNotFound", new Object[]{}, locale);
+        ErrorResponse response = new ErrorResponse();
+        response.setErrorMessage(errorMessage + ' ' + exception.getMessage());
+        response.setErrorCode(exception.getErrorCode());
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/ExceptionController.java]: Let's discuss it on a daily meeting.

File: controller/src/main/java/com/epam/esm/controller/TagController.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * Gets tag.
+     *
+     * @param id the id
+     * @return the tag
+     */
+    @GetMapping(value = "/{id}")
+    public TagDTO getTag(@PathVariable Long id) {
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/TagController.java]:
```suggestion
    public TagDTO findTagById(@PathVariable Long id) {
```

File: model/src/main/java/com/epam/esm/entity/Certificate.java [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+@AllArgsConstructor
+public class Certificate {
+    private Long id;
+    private String name;
+    private String description;
+    private BigDecimal price;
+    private int duration;
+    private ZonedDateTime createDate;
+    private ZonedDateTime lastUpdateDate;
+    Set<Tag> tags;
```
Comment (maximelyashevich) [inline model/src/main/java/com/epam/esm/entity/Certificate.java]:
```suggestion
    private Set<Tag> tags;
```

File: persistence/src/main/java/com/epam/esm/dao/BaseDAO.java [hunk 23]
```diff
@@ -0,0 +1,43 @@
+package com.epam.esm.dao;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * The interface Base dao.
+ *
+ * @param <T> the type parameter
+ * @param <K> the type parameter
```
Comment (maximelyashevich) [inline persistence/src/main/java/com/epam/esm/dao/BaseDAO.java]: Please review `javadoc` for this class.

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+        });
+        String formattedSQL = String.format(SQL_UPDATE_BY_ID, LAST_UPDATE_DATE_COLUMN);
+        jdbcTemplate.update(formattedSQL, ZonedDateTime.now(ZoneId.systemDefault()), id);
+        return true;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        boolean flag;
+        flag = jdbcTemplate.update(SQL_DELETE_CERTIFICATE_BY_ID, id) > 0;
```
Comment (maximelyashevich) [inline persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java]:
```suggestion
        boolean flag = jdbcTemplate.update(SQL_DELETE_CERTIFICATE_BY_ID, id) > 0;
```

File: persistence/src/main/java/com/epam/esm/dao/impl/TagDAOImpl.java [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+                    tagMapper));
+        } catch (EmptyResultDataAccessException e) {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        boolean flag;
+        flag = jdbcTemplate.update(SQL_DELETE_TAG_BY_ID, id) > 0;
```
Comment (maximelyashevich) [inline persistence/src/main/java/com/epam/esm/dao/impl/TagDAOImpl.java]:
```suggestion
        boolean flag = jdbcTemplate.update(SQL_DELETE_TAG_BY_ID, id) > 0;
```

**Review by ykazlovich — CHANGES_REQUESTED** [review]

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java (hunk 1 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java:56]: done!

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java (hunk 2 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: I removed SQLSatatementFormatter, all SQL queries are now in one place(CertificateDAO and TagDAO)

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java (hunk 3 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: done!

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java (hunk 5 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java:39]: done!

File: persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java (hunk 9 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/util/SQLStatementFormatter.java]: done!

File: persistence/src/main/java/com/epam/esm/mapper/CertificateMapper.java (hunk 14 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/mapper/CertificateMapper.java]: done!

File: persistence/src/test/java/com/epam/esm/dao/TagDAOImplTest.java (hunk 15 above)
Comment (virtushawk) [reply persistence/src/test/java/com/epam/esm/dao/TagDAOImplTest.java]: done!

File: controller/build.gradle (hunk 4 above)
Comment (virtushawk) [reply controller/build.gradle]: done!

File: controller/src/main/java/com/epam/esm/config/WebConfig.java (hunk 6 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/config/WebConfig.java]: done!

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 11 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/CertificateController.java]: done!

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 13 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/CertificateController.java]: done!

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 16 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/CertificateController.java]: done!

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 17 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/CertificateController.java]: done!

File: controller/src/main/java/com/epam/esm/controller/ExceptionController.java (hunk 19 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/ExceptionController.java]: done!

File: model/src/main/java/com/epam/esm/entity/Certificate.java (hunk 22 above)
Comment (virtushawk) [reply model/src/main/java/com/epam/esm/entity/Certificate.java]: done!

File: controller/src/main/java/com/epam/esm/controller/TagController.java (hunk 21 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/TagController.java]: what about findTagById?

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java (hunk 24 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java]: done!

File: persistence/src/main/java/com/epam/esm/dao/impl/TagDAOImpl.java (hunk 25 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/dao/impl/TagDAOImpl.java]: done!

File: persistence/src/main/java/com/epam/esm/dao/BaseDAO.java (hunk 23 above)
Comment (virtushawk) [reply persistence/src/main/java/com/epam/esm/dao/BaseDAO.java]: done!

File: controller/src/main/java/com/epam/esm/controller/ExceptionController.java [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+    @ExceptionHandler(InvalidDataFormException.class)
+    public ResponseEntity<Object> handleInvalidDataFormException(InvalidDataFormException exception, Locale locale) {
+        StringBuilder errorMessage = new StringBuilder();
+        errorMessage.append(messageSource.getMessage("error.invalidDataForm", new Object[]{}, locale));
+        for (ObjectError object : exception.getBindingResult().getAllErrors()) {
+            if (object instanceof FieldError) {
+                FieldError fieldError = (FieldError) object;
+                errorMessage.append(messageSource.getMessage(fieldError, locale)).append(',');
+            }
+        }
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/ExceptionController.java]: Can you please rewrite it using `Stream API`?

File: controller/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+    private final CertificateService service;
+
+    /**
+     * finds all certificates by criteria
+     *
+     * @param querySpecificationDTO the query specification dto
+     * @return List of all certificates
+     */
+    @GetMapping
+    public List<CertificateDTO> findAllCertificatesByCriteria(QuerySpecificationDTO querySpecificationDTO) {
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/controller/CertificateController.java]: I would use `findAllByCriteria()` for `CertificateController` class and same for `create()`, `update`, `patchUpdate()` and `delete()` methods. Please review also `TagController` class.

File: persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public boolean delete(Long id) {
+        boolean flag = jdbcTemplate.update(SQL_DELETE_CERTIFICATE_BY_ID, id) > 0;
+        if (flag) {
+            jdbcTemplate.update(SQL_DELETE_TAG_HAS_GIFT_CERTIFICATE_BY_ID, id);
+        }
+        return flag;
+    }
+
+    private String prepareCertificateQueryStatement(QuerySpecification querySpecification) {
```
Comment (maximelyashevich) [inline persistence/src/main/java/com/epam/esm/dao/impl/CertificateDAOImpl.java:132]: You can consider `String.format()` method for `text` and `sort` values.

File: persistence/src/main/java/com/epam/esm/mapper/CertificateMapper.java [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+    @Override
+    public Certificate mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return Certificate.builder().id(rs.getLong("id"))
+                .name(rs.getString("name"))
+                .description(rs.getString("description"))
+                .price(rs.getBigDecimal("price"))
+                .duration(rs.getInt("duration"))
+                .createDate(ZonedDateTime.ofInstant(rs.getTimestamp("create_date").toInstant(), ZoneId.systemDefault()))
+                .lastUpdateDate(ZonedDateTime.ofInstant(rs.getTimestamp("last_update_date").toInstant(), ZoneId.systemDefault()))
+                .build();
```
Comment (maximelyashevich) [inline persistence/src/main/java/com/epam/esm/mapper/CertificateMapper.java]: Please review also `TagMapper` class.
```suggestion
    public Certificate mapRow(ResultSet resultSet, int rowNum) throws SQLException {
        return Certificate.builder()
                .id(resultSet.getLong("id"))
                .name(resultSet.getString("name"))
                .description(resultSet.getString("description"))
                .price(resultSet.getBigDecimal("price"))
                .duration(resultSet.getInt("duration"))
                .createDate(ZonedDateTime.ofInstant(resultSet.getTimestamp("create_date").toInstant(), ZoneId.systemDefault()))
                .lastUpdateDate(ZonedDateTime.ofInstant(resultSet.getTimestamp("last_update_date").toInstant(), ZoneId.systemDefault()))
                .build();
```

File: service/src/main/java/com/epam/esm/service/CertificateService.java [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    CertificateDTO update(CertificateDTO certificateDTO);
+
+    /**
+     * Apply patch certificate dto.
+     *
+     * @param patchDTO the patch dto
+     * @return the certificate dto
+     */
+    CertificateDTO applyPatch(Long id,PatchDTO patchDTO);
```
Comment (maximelyashevich) [inline service/src/main/java/com/epam/esm/service/CertificateService.java]:
```suggestion
    CertificateDTO applyPatch(Long id, PatchDTO patchDTO);
```

File: service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,129 @@
...
+
+    @Override
+    public CertificateDTO findById(Long id) {
+        Optional<Certificate> certificate = certificateDAO.findById(id);
+        if (certificate.isEmpty()) {
+            throw new CertificateNotFoundException(id.toString());
+        }
+        Set<Tag> tags = new HashSet<>(tagDAO.findAllByCertificateId(id));
+        CertificateDTO certificateDTO = mapperDTO.convertCertificateToDTO(certificate.get());
+        Set<TagDTO> tagDTOs = tags.stream().map(mapperDTO::convertTagToDTO).collect(Collectors.toSet());
```
Comment (maximelyashevich) [inline service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: Can we use `distinct()` method from `Steam API`? What do you think?

File: service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+        certificateDTO.setTags(tagDTOs);
+        return certificateDTO;
+    }
+
+    @Override
+    public List<CertificateDTO> findAll(QuerySpecificationDTO querySpecificationDTO) {
```
Comment (maximelyashevich) [inline service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java:73]: To be discussed.

File: controller/src/main/java/com/epam/esm/exception/IdInvalidException.java [hunk 33]
```diff
@@ -0,0 +1,38 @@
+package com.epam.esm.exception;
+
+public class IdInvalidException extends RuntimeException {
+
+    private int errorCode = 101;
```
Comment (maximelyashevich) [inline controller/src/main/java/com/epam/esm/exception/IdInvalidException.java]: Can we declare it as a "constant" field?

**Review by maximelyashevich — CHANGES_REQUESTED** [review]

File: controller/src/main/java/com/epam/esm/controller/ExceptionController.java (hunk 26 above)
Comment (virtushawk) [reply controller/src/main/java/com/epam/esm/controller/ExceptionController.java]: done!

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side.

Regards,
Maks

**Review by ykazlovich — APPROVED** [review]

**Pull request merged**